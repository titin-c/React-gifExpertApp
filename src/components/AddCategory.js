import React, {useState} from 'react';
import PropTypes  from 'prop-types';

export const AddCategory = ({setCategories}) => {

    const [inputValue, setImputValue] = useState('');

    const handleImputChange = (e) => {
        setImputValue(e.target.value);

    }

    const handleSubmit = (e) => {
        //Esto hace que no se refresque la página por defecto al enviar forms
        e.preventDefault();  


        // si la longitud , sin espacios, de valor del impurt es mayor a 0
        if(inputValue.trim().length > 0){
            // Añadimos el valor del input al array de categories
            setCategories( cats => [inputValue, ...cats]); 
            // y vaciamos el inputValue para que no se repita
            setImputValue('');
        }
        
        
    }

  return (
    <div className='buscador'>
        <div>Buscador de GIF's:</div>
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                placeholder='Inserta un Personaje...'
                value={inputValue}
                onChange={handleImputChange}
                >
            </input>
            <button type='submit' >Buscar</button>
        </form>
    </div>
  )
};

/*Es buena practica poner las funciones y valores requeridos en los componentes 
Por si otro programador coge tus componentes... 

a
Aqui indicamos que el componente AddCategory 
necesita obligatoriamente de la función  setCategory
*/

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
 