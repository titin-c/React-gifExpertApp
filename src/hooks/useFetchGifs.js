import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs'


export const useFetchGifs = (category) => {
    
    const [state, setState] = useState({
        data:[],
        loading: true
    });


   
   /*

    Esto hace que cuando se cargue el componente por primera vez, lance la peticiÃ³n 
    para cargar las imagenes

    coge los gifs getGif y los pone en setImages

    Usamos useEffects de react para que no renderice la pÃ¡gina 
    cada vez que note un cambio
    
    */ 
    useEffect(() => {
        
       getGifs( category )
         .then( imgs => {

            setTimeout( ()=>{

             setState({
                 data: imgs,
                 loading: false
             });
            },3000);

         })
            
           


    }, [category]);


    


    return state;
}