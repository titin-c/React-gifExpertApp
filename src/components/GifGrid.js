import React from 'react';
import PropTypes from 'prop-types';

import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({category}) => {

   // llamada al custom hooks useFetchGifs
    
   const {data:images, loading} = useFetchGifs( category );
  
    return (
        <>
            <h2 className='animate__animated animate__fadeInDown'>Imagenes de {category}</h2>

            { loading && <p className='cargando animate__animated animate__pulse'>Cargando... </p>} 
             <div className="card-grid">
                {
                    //mapeamos el array categories para que nos devuelva su contenido
                    //desestructuramos para que nos devuelva el id y el title
                    images.map( img =>(
                    <GifGridItem
                        key={img.id}
                    {...img}
                    />
                        //<li key={id} > {title} </li>
                    ))
                }
            </div> 
            <hr></hr>
        </>
  );
};

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}
