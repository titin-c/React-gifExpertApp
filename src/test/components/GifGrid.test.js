import React  from 'react';
import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Pruebas sobre el componente <GifGrid />', () => {
  
    const category = 'pikatchu';
    test('debe de mostrarse correctamente', () => {

        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow( <GifGrid category={category}/>);

        expect(wrapper).toMatchSnapshot();
       
    });

    test('debe mostrar items cuando se cargan imÃ¡genes useFetchGifs', () => {

        const gifs = [{
            id: 'ABC',
            url: 'https://google.com',
            title: 'prueba de title'
        },
        {
            id: '123',
            url: 'https://google.com',
            title: 'prueba de title'
        }]

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });
      
        const wrapper = shallow( <GifGrid category={category}/>);

        //expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.cargando').exists()).toBe(false);
        expect( wrapper.find('GifGridItem').length).toBe( gifs.length);
        
    });
    
    
    
});
