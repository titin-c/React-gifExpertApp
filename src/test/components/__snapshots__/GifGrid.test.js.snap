// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre el componente <GifGrid /> debe de mostrarse correctamente 1`] = `
<Fragment>
  <hr />
  <h3
    className="animate__animated animate__fadeInDown"
  >
    pikatchu
  </h3>
  <img
    alt="Cargando"
    className="cargando animate__animated animate__pulse"
    src="loading-animation.gif"
  />
  <div
    className="card-grid"
  />
</Fragment>
`;
